### ISSUE TITLE
Evaluating on the complete dev/test dataset by loading a model. Also, maybe some issues with Bert_base_uncased tokenization in Ner class in bert.py ??

### ISSUE THREAD

**Issue by vikas95** [kamalkraj/BERT-NER#57]
Hi @Kamalkraj,

Thanks again for sharing and maintaining this repository frequently. 

For training a system, we have to use - "python run_ner.py --data_dir=data/ --bert_model=bert-base-cased --task_name=ner --output_dir=out_base --max_seq_length=128 --do_train --num_train_epochs 5 --do_eval --warmup_proportion=0.1"

I wanted to ask if we can use a similar command for evaluating a trained model on the entire dev/test files ? Something like -
"python run_ner.py --data_dir=data/ --bert_model=out_base --task_name=ner --max_seq_length=128 --do_eval"

I had tried the example that you have given in Inference block for evaluation (https://github.com/kamalkraj/BERT-NER#inference). I saw before that the tokenization is hardcoded for BERT_cased models in Ner class of bert.py. 
I am currently working with BERT_base_uncased model and during the inference steps, it throws error of mismatch in len(labels) and len(words)

bert.py", line 119, in predict
    assert len(labels) == len(words)
AssertionError

I tried fixing the tokenization in Ner class but then got few other errors. It will be very helpful if you can fix the evaluations for BERT_uncased models. 

Thanks.

**Comment by kamalkraj** [comment]
Hi @vikas95,
use below cmd to run evaluation only :
```
python run_ner.py --data_dir=data/ --bert_model=bert-base-cased --task_name=ner --output_dir=out_base --max_seq_length=128 --do_eval --eval_on=test
```
use `--eval_on` flag to specify `test` or `dev` set to run evalution

**Comment by vikas95** [comment]
Awesome, thanks for the quick response. 

Should I use  --bert_model=bert-large-uncased instead of   --bert_model=bert-base-cased
because I have trained an Bert-large-uncased model ?

**Comment by kamalkraj** [comment]
--bert-model=`the model you used to train`

**Comment by vikas95** [comment]
Hi,

Just wanted to check if you are working on the tokenizer part in Ner class of bert.py

-Vikas

**Comment by kamalkraj** [comment]
what exactly is the issue?

**Comment by vikas95** [comment]
Hi,

Apologies, I had made some changes in bert.py file and the error was coming from my additions. 
Your code part is correct. 

I just had one last question - there is a small difference between perl conlleval.pl evaluation scores and the scores generated by the evaluation codes in this repo. 
I checked the evaluation function but couldn't spot the source of the difference. 

Is it possible to write the output prediction files in the model directory so that we can evaluate it using conll perl script ?

The output prediction file can have conlleval.pl required format:

word1 GOLD_label Predicted_label
word2 GOLD_label Predicted_label
word3 GOLD_label Predicted_label

word1 GOLD_label Predicted_label
word2 GOLD_label Predicted_label
....

Thanks and looking forward for your response..

**Comment by kamalkraj** [comment]
@vikas95,
The max sequence length supported by BERT model is 128. In Dev and Test set there are 2-3 sentence length more than 128, that gets truncated. That is the cause of the small difference in score.

**Comment by vikas95** [comment]
Cool, got it.

Thanks a lot for solving all the doubts and thanks again for sharing+maintaining the codes.

**Comment by vikas95** [comment]
Hi @kamalkraj,

I just had a quick question - can we combine the train and dev sets as a larger training set for training the final model and then evaluate this model on test set ? 

Do we have this feature in the current version of this codes ??

Thanks..

**Comment by kamalkraj** [comment]
@vikas95 
You can combine train and dev set and put it as a train.txt

**Comment by vikas95** [comment]
yep, I was currently doing it this way.

Thanks!
